Error in code generation

The TRACE code differentiation gives another error for clad. There seems to be an error in the intermeditate code generation. I reduced the example to:
```
typedef struct {
} a;
enum { b };
a c;
void d(float) {
  if (b)
    c = (a){};
}
float e;
void f(float) { d(e); }
#include "clad/Differentiator/Differentiator.h"
void g() { clad::gradient(f); }
```

The error output on my side is:
```
test.cpp:4:3: warning: The gradient utilizes a global variable 'c'. Please make sure to properly reset 'c' before re-running the gradient.
    4 | a c;
      |   ^
test.cpp:7:9: warning: attempted to differentiate unsupported statement, no changes applied
    7 |     c = (a){};
      |         ^
test.cpp:9:7: warning: The gradient utilizes a global variable 'e'. Please make sure to properly reset 'e' before re-running the gradient.
    9 | float e;
      |       ^
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /home/msagebaum/Applications/llvm-project/install/bin/clang++ -fplugin=/home/msagebaum/Applications/clad/inst_debug/lib/clad.so -I/home/msagebaum/Applications/clad/inst_debug/include -Xclang -plugin-arg-clad -Xclang -fgenerate-source-file -I/home/msagebaum/Kaiserslautern/mtu/Programms/trace_suite/trace/INCLUDE -I/home/msagebaum/Kaiserslautern/mtu/Programms/trace_suite/build_jacobianIndex_dbg/trace -I/home/msagebaum/Applications/build/include -I/usr/include/openmpi-x86_64/ -I/home/msagebaum/Kaiserslautern/mtu/Programms/trace_suite/trace/AD -I/home/msagebaum/Kaiserslautern/mtu/Programms/trace_suite/contrib/ad_dependencies/adToolbox/include test.cpp -c -o test.o -DNDEBUG -DSIMD_STRUCTS
1.	<eof> parser at end of file
2.	LLVM IR generation of declaration 'd_pullback'
3.	Generating code for declaration 'd_pullback'
4.	test.cpp:1:1: LLVM IR generation of compound statement ('{}')
 #0 0x0000000003567f68 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x3567f68)
 #1 0x0000000003565d44 llvm::sys::CleanupOnSignal(unsigned long) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x3565d44)
 #2 0x00000000034a6398 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x00007ff501428070 __restore_rt (/lib64/libc.so.6+0x1a070)
 #4 0x00000000038c963c (anonymous namespace)::AggExprEmitter::VisitCXXParenListOrInitListExpr(clang::Expr*, llvm::ArrayRef<clang::Expr*>, clang::FieldDecl*, clang::Expr*) CGExprAgg.cpp:0:0
 #5 0x00000000038cc976 clang::CodeGen::CodeGenFunction::EmitAggExpr(clang::Expr const*, clang::CodeGen::AggValueSlot) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x38cc976)
 #6 0x00000000038cc00c clang::StmtVisitorBase<std::add_pointer, (anonymous namespace)::AggExprEmitter, void>::Visit(clang::Stmt*) CGExprAgg.cpp:0:0
 #7 0x00000000038cc976 clang::CodeGen::CodeGenFunction::EmitAggExpr(clang::Expr const*, clang::CodeGen::AggValueSlot) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x38cc976)
 #8 0x000000000389917b clang::CodeGen::CodeGenFunction::EmitAnyExprToMem(clang::Expr const*, clang::CodeGen::Address, clang::Qualifiers, bool) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x389917b)
 #9 0x00000000038a7b74 clang::CodeGen::CodeGenFunction::EmitMaterializeTemporaryExpr(clang::MaterializeTemporaryExpr const*) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x38a7b74)
#10 0x00000000038bcb6b clang::CodeGen::CodeGenFunction::EmitLValueHelper(clang::Expr const*, clang::CodeGen::KnownNonNull_t) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x38bcb6b)
#11 0x00000000038bd644 void llvm::function_ref<void ()>::callback_fn<clang::CodeGen::CodeGenFunction::EmitLValue(clang::Expr const*, clang::CodeGen::KnownNonNull_t)::'lambda'()>(long) CGExpr.cpp:0:0
#12 0x00000000072cd992 clang::StackExhaustionHandler::runWithSufficientStackSpace(clang::SourceLocation, llvm::function_ref<void ()>) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x72cd992)
#13 0x00000000038ba68e clang::CodeGen::CodeGenFunction::EmitReferenceBindingToExpr(clang::Expr const*) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x38ba68e)
#14 0x0000000003e0058b clang::CodeGen::CodeGenFunction::EmitCallArg(clang::CodeGen::CallArgList&, clang::Expr const*, clang::QualType) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x3e0058b)
#15 0x0000000003e01abe clang::CodeGen::CodeGenFunction::EmitCallArgs(clang::CodeGen::CallArgList&, clang::CodeGen::CodeGenFunction::PrototypeWrapper, llvm::iterator_range<clang::Stmt::CastIterator<clang::Expr, clang::Expr const* const, clang::Stmt const* const>>, clang::CodeGen::CodeGenFunction::AbstractCallee, unsigned int, clang::CodeGen::CodeGenFunction::EvaluationOrder) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x3e01abe)
#16 0x00000000038d2927 commonEmitCXXMemberOrOperatorCall(clang::CodeGen::CodeGenFunction&, clang::GlobalDecl, llvm::Value*, llvm::Value*, clang::QualType, clang::CallExpr const*, clang::CodeGen::CallArgList&, clang::CodeGen::CallArgList*) CGExprCXX.cpp:0:0
#17 0x00000000038d2fa3 clang::CodeGen::CodeGenFunction::EmitCXXMemberOrOperatorCall(clang::CXXMethodDecl const*, clang::CodeGen::CGCallee const&, clang::CodeGen::ReturnValueSlot, llvm::Value*, llvm::Value*, clang::QualType, clang::CallExpr const*, clang::CodeGen::CallArgList*, llvm::CallBase**) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x38d2fa3)
#18 0x00000000038deaa8 clang::CodeGen::CodeGenFunction::EmitCXXMemberOrOperatorMemberCallExpr(clang::CallExpr const*, clang::CXXMethodDecl const*, clang::CodeGen::ReturnValueSlot, bool, clang::NestedNameSpecifier*, bool, clang::Expr const*, llvm::CallBase**) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x38deaa8)
#19 0x00000000038df38a clang::CodeGen::CodeGenFunction::EmitCXXMemberCallExpr(clang::CXXMemberCallExpr const*, clang::CodeGen::ReturnValueSlot, llvm::CallBase**) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x38df38a)
#20 0x00000000038ad81b clang::CodeGen::CodeGenFunction::EmitCallExpr(clang::CallExpr const*, clang::CodeGen::ReturnValueSlot, llvm::CallBase**) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x38ad81b)
#21 0x00000000038bd966 clang::CodeGen::RValue llvm::function_ref<clang::CodeGen::RValue (clang::CodeGen::ReturnValueSlot)>::callback_fn<(anonymous namespace)::AggExprEmitter::VisitCallExpr(clang::CallExpr const*)::'lambda'(clang::CodeGen::ReturnValueSlot)>(long, clang::CodeGen::ReturnValueSlot) CGExprAgg.cpp:0:0
#22 0x00000000038c83d0 (anonymous namespace)::AggExprEmitter::withReturnValueSlot(clang::Expr const*, llvm::function_ref<clang::CodeGen::RValue (clang::CodeGen::ReturnValueSlot)>) CGExprAgg.cpp:0:0
#23 0x00000000038c883f (anonymous namespace)::AggExprEmitter::VisitCallExpr(clang::CallExpr const*) CGExprAgg.cpp:0:0
#24 0x00000000038cc976 clang::CodeGen::CodeGenFunction::EmitAggExpr(clang::Expr const*, clang::CodeGen::AggValueSlot) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x38cc976)
#25 0x00000000038a43a2 clang::CodeGen::CodeGenFunction::EmitAnyExpr(clang::Expr const*, clang::CodeGen::AggValueSlot, bool) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x38a43a2)
#26 0x00000000038a6a4c clang::CodeGen::CodeGenFunction::EmitIgnoredExpr(clang::Expr const*) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x38a6a4c)
#27 0x00000000039fcaa2 clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x39fcaa2)
#28 0x0000000003a0307b clang::CodeGen::CodeGenFunction::EmitIfStmt(clang::IfStmt const&) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x3a0307b)
#29 0x0000000003a03889 clang::CodeGen::CodeGenFunction::EmitCompoundStmtWithoutScope(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x3a03889)
#30 0x0000000003a03c3d clang::CodeGen::CodeGenFunction::EmitCompoundStmt(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x3a03c3d)
#31 0x0000000003a03ec5 clang::CodeGen::CodeGenFunction::EmitSimpleStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x3a03ec5)
#32 0x00000000039fca00 clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x39fca00)
#33 0x0000000003a03889 clang::CodeGen::CodeGenFunction::EmitCompoundStmtWithoutScope(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x3a03889)
#34 0x0000000003a69c12 clang::CodeGen::CodeGenFunction::EmitFunctionBody(clang::Stmt const*) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x3a69c12)
#35 0x0000000003a7e526 clang::CodeGen::CodeGenFunction::GenerateCode(clang::GlobalDecl, llvm::Function*, clang::CodeGen::CGFunctionInfo const&) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x3a7e526)
#36 0x0000000003ace067 clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x3ace067)
#37 0x0000000003ac969d clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x3ac969d)
#38 0x0000000003ac9e23 clang::CodeGen::CodeGenModule::EmitGlobal(clang::GlobalDecl) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x3ac9e23)
#39 0x0000000003ad2b5b clang::CodeGen::CodeGenModule::EmitTopLevelDecl(clang::Decl*) (.part.0) CodeGenModule.cpp:0:0
#40 0x0000000003efc4b0 (anonymous namespace)::CodeGeneratorImpl::HandleTopLevelDecl(clang::DeclGroupRef) ModuleBuilder.cpp:0:0
#41 0x0000000003ee7600 clang::BackendConsumer::HandleTopLevelDecl(clang::DeclGroupRef) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x3ee7600)
#42 0x000000000425e8f0 clang::MultiplexConsumer::HandleTopLevelDecl(clang::DeclGroupRef) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x425e8f0)
#43 0x00007ff4ff61e89f clad::plugin::CladPlugin::SendToMultiplexer() /home/msagebaum/Applications/clad/tools/ClangPlugin.cpp:364:11
#44 0x00007ff4ff61ef61 clad::plugin::CladPlugin::HandleTranslationUnit(clang::ASTContext&) /home/msagebaum/Applications/clad/tools/ClangPlugin.cpp:527:7
#45 0x000000000425ea00 clang::MultiplexConsumer::HandleTranslationUnit(clang::ASTContext&) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x425ea00)
#46 0x0000000005aacf3c clang::ParseAST(clang::Sema&, bool, bool) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x5aacf3c)
#47 0x00000000042170a1 clang::FrontendAction::Execute() (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x42170a1)
#48 0x0000000004195ec2 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x4195ec2)
#49 0x00000000042f1081 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x42f1081)
#50 0x00000000005a7a08 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x5a7a08)
#51 0x000000000059f214 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#52 0x0000000003f4b669 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::'lambda'()>(long) Job.cpp:0:0
#53 0x00000000034a6754 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x34a6754)
#54 0x0000000003f4bd30 clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (.part.0) Job.cpp:0:0
#55 0x0000000003f0bf59 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x3f0bf59)
#56 0x0000000003f0cee2 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x3f0cee2)
#57 0x0000000003f224bc clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x3f224bc)
#58 0x00000000005a3b95 clang_main(int, char**, llvm::ToolContext const&) (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x5a3b95)
#59 0x00000000004508ef main (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x4508ef)
#60 0x00007ff501411575 __libc_start_call_main (/lib64/libc.so.6+0x3575)
#61 0x00007ff501411628 __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x3628)
#62 0x000000000059ec55 _start (/home/msagebaum/Applications/llvm-project/install/bin/clang+++0x59ec55)
clang++: error: clang frontend command failed with exit code 139 (use -v to see invocation)
clang version 20.1.7 (https://github.com/llvm/llvm-project.git 6146a88f60492b520a36f8f8f3231e15f3cc6082)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/msagebaum/Applications/llvm-project/install/bin
clang++: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang++: note: diagnostic msg: /tmp/test-924a15.cpp
clang++: note: diagnostic msg: /tmp/test-924a15.sh
clang++: note: diagnostic msg: 

********************
```

The command line is:
```
clang++ -fplugin=clad.so -Iclad/include -Xclang -plugin-arg-clad -Xclang -fgenerate-source-file test.cpp -c -o test.o
```

--- a/include/clad/Differentiator/ReverseModeVisitor.h
+++ b/include/clad/Differentiator/ReverseModeVisitor.h
@@ -371,6 +371,7 @@ namespace clad {
     StmtDiff VisitCallExpr(const clang::CallExpr* CE);
     virtual StmtDiff VisitCompoundStmt(const clang::CompoundStmt* CS);
     StmtDiff VisitConditionalOperator(const clang::ConditionalOperator* CO);
+    StmtDiff VisitCompoundLiteralExpr(const clang::CompoundLiteralExpr* CLE);
     StmtDiff VisitCXXBoolLiteralExpr(const clang::CXXBoolLiteralExpr* BL);
     StmtDiff VisitCXXBindTemporaryExpr(const clang::CXXBindTemporaryExpr* BTE);
     StmtDiff VisitCharacterLiteral(const clang::CharacterLiteral* CL);

--- a/lib/Differentiator/ReverseModeVisitor.cpp
+++ b/lib/Differentiator/ReverseModeVisitor.cpp
@@ -714,6 +714,18 @@ Expr* ReverseModeVisitor::getStdInitListSizeExpr(const Expr* E) {
     return StmtDiff(Clone(S));
   }
 
+  StmtDiff
+  ReverseModeVisitor::VisitCompoundLiteralExpr(const CompoundLiteralExpr* CLE) {
+    StmtDiff result = Visit(CLE->getInitializer());
+    ParsedType PT = ParsedType::make(CLE->getType());
+    result.updateStmt(
+        m_Sema.ActOnCompoundLiteral(noLoc, PT, noLoc, result.getExpr()).get());
+    result.updateStmtDx(
+        m_Sema.ActOnCompoundLiteral(noLoc, PT, noLoc, result.getExpr_dx())
+            .get());
+    return result;
+  }
+
   StmtDiff ReverseModeVisitor::VisitCompoundStmt(const CompoundStmt* CS) {
     int scopeFlags = Scope::DeclScope;
     // If this is the outermost compound statement of the function,

--- a/test/Gradient/UserDefinedTypes.C
+++ b/test/Gradient/UserDefinedTypes.C
@@ -1285,6 +1285,17 @@ double fn34(double x, double y) {
 // CHECK-NEXT:    *_d_x += _d_obj_x.data;
 // CHECK-NEXT:}
 
+double fn35(double x, double y) {
+    double& ref = *(PtrAndValAggr){x, &y}.ptr;
+    return ref;
+}
+
+// CHECK:  void fn35_grad(double x, double y, double *_d_x, double *_d_y) {
+// CHECK-NEXT:      double &_d_ref = *(PtrAndValAggr){0., _d_y}.ptr;
+// CHECK-NEXT:      double &ref = *(PtrAndValAggr){x, &y}.ptr;
+// CHECK-NEXT:      _d_ref += 1;
+// CHECK-NEXT:  }
+
 void print(const Tangent& t) {
   for (int i = 0; i < 5; ++i) {
     printf("%.2f", t.data[i]);
@@ -1422,4 +1433,7 @@ int main() {
 
     INIT_GRADIENT(fn34);
     TEST_GRADIENT(fn34, /*numOfDerivativeArgs=*/2, -5, 6, &d_i, &d_j);    // CHECK-EXEC: {1.00, 2.00}
+
+    INIT_GRADIENT(fn35);
+    TEST_GRADIENT(fn35, /*numOfDerivativeArgs=*/2, -5, 6, &d_i, &d_j);    // CHECK-EXEC: {0.00, 1.00}
 }

